### PR TITLE
Adding a scheme for encrypting additive shares

### PR DESCRIPTION
**Pull request by martinthomson**
XOR shares are just addition in a group where `|G| = 2` so adding those
will be easy enough.

### CODE REVIEW

**Review by benjaminsavage — APPROVED** [review]
I love it. Simple, elegant.

The key encryption / decryption logic could benefit from some comments.

**Review by martinthomson — COMMENTED** [review]
Simple, elegant, and wrong.

One reason we're using (partially) homomorphic encryption here is to prevent the client from sending information to helper1 or helper2.

This protects the ciphertext, but does not protect the secret share, giving clients a channel by which they can signal to helper3/helper4.  The client can just tell helper3 which "secret" value they used.

I haven't been able to come up with anything better than Paillier on a per-value basis.  We might be able to use Pollard's Rho algorithm ([paper](https://cryptojedi.org/papers/negation-20110102.pdf)) and use the same EC ElGamal as used here, but it would be per-value protection, so 64 bytes per value.  The value space would have to be more limited as well, as that algorithm runs in O(n^{1/2}), which could quickly get out of hand.

There might be a point at which we have something functional, but it is unlikely we will find something that will work with Prio as simple operations there operate in a prime group that is at least as large as the largest sum, divided by the probability that cheating is successful.  For us, if we imagine batching ~2^14 events, with values of each up to 2^16, and a cheating probability of ~2^{-64}, that pushes the group toward 2^100.  That's 2^50 operations *per value*, with 2^4 values.

The paper I cited earlier dedicated an entire machine (a 2011 machine, sure) to the task for a much smaller curve.  That resulted in producing less than 50 candidate collisions per second, with a subgroup of size 2^50 resolving in around half a second, a 2^55 subgroup taking 1.5 seconds, and a 2^60 subgroup taking about 16 seconds.  Those are medians; maximums for the last was 72 seconds of hard work.  Even with 10 years of raw speed improvements, 2^100 is still likely out of reach.

Even if we were to dispense with Prio and we constrained the value space aggressively, it might not come down to a size and complexity that is better than Paillier.  Furthermore, I was not able to find any implementation of Pollard's Rho that would suit us.  They exist, but (there are some in multiplicative prime groups as far as I can see).

File: raw-ipa-lib/src/ss.rs [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+from_smaller!(u32, AdditiveShare<32>);
+from_smaller!(u16, AdditiveShare<16>);
+from_smaller!(u8, AdditiveShare<8>);
+
+impl<I, const N: u32> Add<I> for AdditiveShare<N>
+where
+    I: Into<u128>,
+{
+    type Output = Self;
+    fn add(mut self, rhs: I) -> Self::Output {
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply raw-ipa-lib/src/ss.rs:89]: So it's returning `AdditiveShare<const N: u32>:: AdditiveShare<const N: u32>`? I don't get it.

File: raw-ipa-lib/src/ss.rs [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+        self.v = self.v.wrapping_sub(rhs.into());
+        self.wrap();
+    }
+}
+
+impl<const N: u32> AddAssign<Self> for AdditiveShare<N> {
+    fn add_assign(&mut self, rhs: Self) {
+        self.v += rhs.v;
+    }
+}
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply raw-ipa-lib/src/ss.rs:144]: OK, I'd recommend using the "more magic" approach then if it removes the need to duplicate the code.

File: raw-ipa-lib/src/ss.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinthomson) [reply raw-ipa-lib/src/ss.rs:89]: `type` creates a type alias, just like `typedef` in C.  So this is returning `AdditiveShare<const N: u32>`, spelled differently.  One important thing to note is that the `N` is the same `N` as in the `impl` declaration, so you can't go from `AdditiveShare<32>` to `AdditiveShare<64>`.

**Comment by martinthomson** [comment]
Abandoning this one.

**Pull request closed without merging**